chain/gap/find.go

```diff
@@ -0,0 +1,240 @@
+package gap
+
+import (
+	"context"
+	"time"
+
+	"github.com/deckarep/golang-set"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/sentinel-visor/chain"
+	"github.com/filecoin-project/sentinel-visor/lens"
+	"github.com/filecoin-project/sentinel-visor/model/visor"
+	"github.com/filecoin-project/sentinel-visor/storage"
+	"golang.org/x/xerrors"
+)
+
+type GapIndexer struct {
+	DB                   *storage.Database
+	opener               lens.APIOpener
+	name                 string
+	minHeight, maxHeight uint64
+}
+
+var TaskSet mapset.Set
+
+func init() {
+	TaskSet = mapset.NewSet()
+	for _, t := range chain.AllTasks {
+		TaskSet.Add(t)
+	}
+}
+
+func NewGapIndexer(o lens.APIOpener, db *storage.Database, name string, maxHeight, minHeight uint64) *GapIndexer {
+	return &GapIndexer{
+		DB:        db,
+		opener:    o,
+		name:      name,
+		maxHeight: maxHeight,
+		minHeight: minHeight,
+	}
+}
+
+func (g *GapIndexer) Run(ctx context.Context) error {
+	startTime := time.Now()
+	node, closer, err := g.opener.Open(ctx)
+	if err != nil {
+		return xerrors.Errorf("open lens: %w", err)
+	}
+	defer func() {
+		closer()
+	}()
+
+	head, err := node.ChainHead(ctx)
+	if err != nil {
+		return err
+	}
+	maxHeight := g.maxHeight
+	if uint64(head.Height()) < g.maxHeight {
+		maxHeight = uint64(head.Height())
+	}
+
+	findLog := log.With("type", "find")
+
+	taskGaps, err := g.findTaskEpochGaps(ctx, len(chain.AllTasks), maxHeight, g.minHeight)
+	if err != nil {
+		return xerrors.Errorf("finding task epoch gaps: %w", err)
+	}
+	findLog.Infow("found gaps in tasks", "count", len(taskGaps))
+
+	heightGaps, nulls, err := g.findEpochGapsAndNullRounds(ctx, node, maxHeight, g.minHeight)
+	if err != nil {
+		return xerrors.Errorf("finding epoch gaps: %w", err)
+	}
+	findLog.Infow("found gaps in epochs", "count", len(heightGaps))
+	findLog.Infow("found null rounds", "count", len(nulls))
+
+	skipGaps, err := g.findEpochSkips(ctx, maxHeight, g.minHeight)
+	if err != nil {
+		return xerrors.Errorf("detecting skipped gaps: %w", err)
+	}
+	findLog.Infow("found skipped epochs", "count", len(skipGaps))
+
+	var nullRounds visor.ProcessingReportList
+	for _, epoch := range nulls {
+		nullRounds = append(nullRounds, &visor.ProcessingReport{
+			Height:      int64(epoch),
+			StateRoot:   "NULL_ROUND",
+			Reporter:    g.name,
+			Task:        "NULL_ROUND",
+			StartedAt:   startTime,
+			CompletedAt: time.Now(),
+			Status:      visor.ProcessingStatusOK,
+		})
+	}
+
+	return g.DB.PersistBatch(ctx, skipGaps, heightGaps, taskGaps, nullRounds)
+}
+
+type GapIndexerLens interface {
+	ChainGetTipSetByHeight(ctx context.Context, epoch abi.ChainEpoch, tsk types.TipSetKey) (*types.TipSet, error)
+}
+
+func (g *GapIndexer) findEpochSkips(ctx context.Context, max, min uint64) (visor.GapReportList, error) {
```

Comment:
Can you split this into a separate type and introduce a `GapFinder` interface? I'd like to be able to reuse some of this logic by passing in a different implementation (reading from csv dump for example).

Reply:
Yes, what would you like it to look like? Maybe:
```go
type GapFinder interface {
    Find(ctx context.Context, to, from uint64, tasks ...string) (visor.GapReportList, []abi.ChainEpoch, error)
}
``` 

The returned array `[]abi.ChainEpoch` containing null rounds

Reply:
On second thoughts I don't think it adds much. We can simply replicate logic if needed - it's pretty short.